File: client/lib/connection/impl/updateconvospacket.dart
import 'package:client/connection/connection.dart';
import 'dart:typed_data';

import 'package:client/connection/packet.dart';
import 'package:client/conversation/conversation.dart';
import 'package:client/utils/bytebuffer.dart';

class UpdateConvosPacket extends Packet {
  final List<Conversation> conversations = [];

  UpdateConvosPacket.write();
  UpdateConvosPacket.read();

  @override
  int getId() {
    return 4;
  }

  @override
  Uint8List write() {
    return Uint8List(0);
  }

  @override
  void read(Uint8List data) {
    conversations.clear();
    ByteBuf buf = ByteBuf.read(data);
    int size = buf.readInt();
    for (int i = 0; i < size; i++) {
      String name = buf.readString();
      bool b = buf.readBoolean();
      String? icon;
      if (b) {
        icon = buf.readString();
      }
      int usize = buf.readInt();
      List<String> users = [];
      for (int i = 0; i < usize; i++) {
        users.add(buf.readString());
      }
      conversations.add(Conversation(name, icon, users));
    }
  }

  @override
  void handle(Connection connection) {
    connection.eventBus.onUpdateConvos(connection, conversations);
  }
}

File: server/Server/src/me/spencernold/server/connection/packet/impl/UpdateConvosPacket.java
package me.spencernold.server.connection.packet.impl;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import me.spencernold.server.connection.Client;
import me.spencernold.server.connection.packet.Id;
import me.spencernold.server.connection.packet.Packet;
import me.spencernold.server.messages.Conversation;
import me.spencernold.server.messages.UserConvoHeader;
import me.spencernold.server.util.ByteBuffer;

@Id(id = 4)
public class UpdateConvosPacket implements Packet {

	private final List<Conversation> conversations;
	
	public UpdateConvosPacket(List<Conversation> conversations) {
		this.conversations = conversations;
	}
	
	public UpdateConvosPacket() {
		conversations = null;
	}
	
	public byte[] write() {
		ByteBuffer buf = new ByteBuffer();
		buf.writeInt(conversations.size());
		for (Conversation c : conversations) {
			buf.writeString(c.getName());
			buf.writeOptionalString(c.getIcon());
			List<String> users = c.getUsers();
			buf.writeInt(users.size());
			for (String s : users)
				buf.writeString(s);
		}
		return buf.toByteArray();
	}

	public void read(ByteBuffer buf) throws Exception {
	}

	public void handle(Client client) {
		if (client.user == null)
			return;
		try {
			List<Conversation> conversations = new ArrayList<>();
			UserConvoHeader header = client.server.data.getUserConvoHeader();
			for (byte[] secretKey : header.getUserConvos(client.user.getName())) {
				Conversation c = Conversation.loadConvo(secretKey);
				conversations.add(c);
			}
			client.send(new UpdateConvosPacket(conversations));
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}
